sphinx/builders/__init__.py

```diff
@@ -581,11 +612,34 @@ def _write_serial(self, docnames: Sequence[str]) -> None:
                 self.write_doc(docname, doctree)
 
     def _write_parallel(self, docnames: Sequence[str], nproc: int) -> None:
-        def write_process(docs: list[tuple[str, nodes.document]]) -> None:
+        def write_process_serial_post_transform(
+            docs: list[tuple[str, nodes.document]],
+        ) -> None:
             self.app.phase = BuildPhase.WRITING
+            # The doctree has been post-transformed (incl. write_doc_serialized)
+            # in the main process, only write_doc() is needed here
             for docname, doctree in docs:
                 self.write_doc(docname, doctree)
 
+        def write_process_parallel_post_transform(docs: list[str]) -> bytes:
+            assert self.env.config.enable_parallel_post_transform
+            self.app.phase = BuildPhase.WRITING
+            # run post-transform, doctree-resolved and write_doc_serialized in parallel
+            for docname in docs:
+                doctree = self.env.get_and_resolve_doctree(docname, self)
+                # write_doc_serialized is assumed to be safe for all Sphinx
+                # internal builders. Some builders merge information from post-transform
+                # and write_doc_serialized back to the main process using
+                # Builder.post_transform_merge_attr and
+                # Builder.merge_builder_post_transform
+                self.write_doc_serialized(docname, doctree)
+                self.write_doc(docname, doctree)
+            merge_attr = {
```

Comment:
This is not a robust enough mechanism for merging data generated during post-processing:

- It only accounts for data saved on to the `Builder`, not on to the `Environment`
- It doesn't give any hook for extensions to merge data

Reply:
There should be two clear "hook" points, that "anyone" (builders, extensions, etc) can use:

1. A hook function/method where they can write logic to add data to the "entity" that will be pickled and returned from a single process
2. A hook function/method where they can write logic to "merge" the pickled data back into the core process